+bids/layout.m

```diff
@@ -456,66 +453,120 @@ function tolerant_message(use_schema, msg)
 
 function structure = manage_tsv(structure, pth, filename)
 
-  p = bids.internal.file_utils('FPList', pth,  ['^' strrep(filename, '.tsv', '\.tsv') '$']);
+  ext = bids.internal.file_utils(filename, 'ext');
+  tsv_file = bids.internal.file_utils('FPList', ...
+                                      pth,  ...
+                                      ['^' strrep(filename, ['.' ext], ['\.' ext]) '$']);
 
-  if isempty(p)
+  if isempty(tsv_file)
     warning('Missing: %s', fullfile(pth, filename));
 
   else
-    structure.content = bids.util.tsvread(p);
+    structure.content = bids.util.tsvread(tsv_file);
+
+    tsv_file = bids.internal.file_utils('FPList', ...
+                                        pth,  ...
+                                        ['^' strrep(filename, ['.' ext], '\.json') '$']);
+    if ~isempty(tsv_file)
+      structure.meta = bids.util.jsondecode(tsv_file);
```

Comment:
If both tsv and json are readed into memory, it would be worth to check their consistency -- ie if all columns in tsv are defined in json.

Reply:
Yeah not sure about especially if we drop reading the files during indexing (see below).

Reply:
It is question of design. My opinion (purely subjectif):
For layout metadata is needed only for parcing IntendedFor fields
For query returning full structure and just path to json file are equivalent, retrieving full stucture is trivial using `bids.util.jsondecode`, hence no need to return full structure.
But, what can be useful is to query on individual fields of meta, i.e. "show me all files with repetition time == 0.03". Then contains of meta will become important for query. But it is difficult to implement (espetualy if queried values are numbers and not strings).

The question if load in memory all meta, or will read given file each time we look in it -- is open. One economise time, other memory. For me it is better to read each time -- layout and query will be called once, and before processing images which will dominate calculation time. The json files are small, so user will see slow-down only in huge datasets (but there any solution will be slow).

For indexing, I can't comment beacause of lack of matlab knowelege

Reply:
> It is question of design. My opinion (purely subjectif):
:wink: 

> For layout metadata is needed only for parcing IntendedFor fields
> For query returning full structure and just path to json file are equivalent, retrieving full stucture is trivial using `bids.util.jsondecode`, hence no need to return full structure.

Better to use `get_metadata` that should take care of the inheritance principle but I see what you mean.

> But, what can be useful is to query on individual fields of meta, i.e. "show me all files with repetition time == 0.03". Then contains of meta will become important for query. But it is difficult to implement (espetualy if queried values are numbers and not strings).

There is some code "ruins" in `bids.query` about this: so I will leave that there for though I agree that doing this on query could be painful.

> The question if load in memory all meta, or will read given file each time we look in it -- is open. One economise time, other memory. For me it is better to read each time -- layout and query will be called once, and before processing images which will dominate calculation time. The json files are small, so user will see slow-down only in huge datasets (but there any solution will be slow).

I agree that the bottle neck in our pipeline is definitely not json metadata reading. And when moving to big data set, we might have to go a completely different way anyway.